lib/supavisor_web/ws_proxy.ex

```diff
@@ -0,0 +1,78 @@
+defmodule SupavisorWeb.WsProxy do
+  @moduledoc """
+  The `Supavisor.WsProxy` module implements a WebSocket proxy for managing
+  TCP connections between clients and a Postgres database.
+  """
+
+  require Logger
+  alias Supavisor.Protocol.Server
+  @behaviour Phoenix.Socket.Transport
+
+  def child_spec(_opts) do
+    %{
+      id: Task,
+      start: {Task, :start_link, [fn -> :ok end]},
+      restart: :transient
+    }
+  end
+
+  def connect(map), do: {:ok, map}
+
+  def init(_) do
+    proxy_port = Application.fetch_env!(:supavisor, :proxy_port)
+
+    {:ok, socket} =
+      :gen_tcp.connect('localhost', proxy_port, [:binary, packet: :raw, active: true])
+
+    {:ok, %{status: :startup, acc: "", socket: socket}}
+  end
+
+  def handle_in(
+        {<<len::32, startup_pkt::binary-size(len - 4), rest::binary>>, _},
+        %{status: :startup} = state
+      ) do
+    :ok = :gen_tcp.send(state.socket, [<<len::32>>, startup_pkt])
+
+    {:ok, %{state | acc: rest}}
+  end
+
+  def handle_in({bin, _opts}, %{socket: socket} = state) do
+    :ok = :gen_tcp.send(socket, bin)
+    {:ok, state}
+  end
+
+  def handle_info({:tcp, _, bin}, %{status: :startup} = state) do
+    if String.ends_with?(bin, Server.ready_for_query()) do
+      acc = filter_pass_pkt(state.acc)
+      :ok = :gen_tcp.send(state.socket, acc)
```

Comment:
FWIW, this may return `{:error, _}`, but letting it crash is fine (and probably what would be done anyway).